### ISSUE TITLE
Random data augmentation in segmentation pipeline? Call for an official example 

### ISSUE THREAD

**Issue by Beanocean** [NVIDIA/DALI#1138]
I am building a data augmentation pipeline on image scene segmentation with DALI.

In my situation, I need to do some synchronized data augmentations both to the source image and the target label image. The main procedures are listed in the pseudo-code below.

``` python
image = ops.ImageDecoder(image, output_type=types.RGB)
label = ops.ImageDecoder(label, output_type=types.GRAY)

# Scale the image relativly, still a open problem, 
# ref: [Issue: #969](https://github.com/NVIDIA/DALI/issues/969)
# Requirments: relative random scalling, scale range: [0.5, 2.0], step size: 0.25
# DALI seems only support absolute coordinate 
image, label = ops.Resize([image, label], shorter_size=ops.Uniform([image_width*0.5, image_width*2.0]))

# Requirements: crop the images to a fixed size
# Error will occur when the image size is smaller than the target cropping size.
# How to make a workaround? I find the `ops.Paste`, but it's not suitable to apply to the enlarged images.
image, label = ops.Crop([image, label], crop_w=crop_size[0], crop_h=crop_size[1])

# Random mirror
image, label = ops.Flip([image, label], horizontal=ops.CoinFlip(probability=0.5))

# Convert image, label into NCHW layouts, without using the mirroring and cropping operations.
# A standard normalization operation is required as well as a cast operation converting tensors into a fixed layout.
image = ops.CropMirrorNormalize(mean=[128, 128, 128], std=[1,1,1,], output_layout=types.NCHW)
label = ops.CropMirrorNormalize(mean=[0, 0, 0], std=[1,1,1,], output_layout=types.NCHW)
```

I have found some workarounds to meet some requirements above, but there is still an obstacle to crop the image whose size is smaller than the target cropping size.

Any response will be appreciated.

Regards,

**Comment by JanuszL** [comment]
Hi @Beanocean,

Regarding:
* Crop - there is some work to make padding possible (with the value set) inside crop - https://github.com/NVIDIA/DALI/pull/1129. What is left to be done is to adjust the operator itself which should follow it not too far from now.
* Resize - for that we need to have the arithmetic operators working on the tensors. This is still in our backlog, but with quite high priority. Then it should be possible to create op that extracts the sizes from tensors and do something like:
```
rand = ops.Uniform(range = (0.5, 2.))
sizes = ops.GetSizes()
res = ops.Resize()
(...)
scales = rand * sizes
image, label = ops.Resize([image, label], shorter_size=scales)
```

**Comment by Beanocean** [comment]
@JanuszL Cool! Thanks for doing such an awesome project.

Just sharing my code clip here as an example. I only test the code on Jupyter notebook and the displayed images and masks look OK.

The correctness on the training procedure still remains to be tested !!!

<script src="https://gist.github.com/Beanocean/62ba8de2deb5bb2d76474cef2a22bb61.js"></script>

**Comment by JanuszL** [comment]
BTW - if there is some op missing you can still implement it in python  - check this [tutorial](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/python_operator.html).  The performance would be inferior to the native implementation but it should give you the flexibility to experiment with different operations.

**Comment by Beanocean** [comment]
Thanks for your advice. I have read the tutorial about python operator. It's a good tool, but the data transfer between GPU and CPU is expensive. Is that possible to use `pytorch.cuda.opeartions` in the python operator?

**Comment by JanuszL** [comment]
It is not possible now, you can only execute python code on the CPU.
But we are exploring the possibility to support this kind of operations on GPU as well.
@banasraf ?

**Comment by banasraf** [comment]
Hi @Beanocean 
I've started investigating possibilities of implementing GPU `PythonFunction`. It's hard to guess when such feature could arrive because before that there is some preparation work to be done. Probably we'll go for exposing DALI Tensors to Python as DLpack Tensors, which give nice flexibility as those can be consumed, e.g. by CuPy arrays or by torch tensors. 
Also, lately `TorchPythonFunction` was implemented, which is a wrapper for `PythonFunction` that runs a function with torch tensors as arguments. When the GPU support arrives, this operator will probably be capable of executing functions working on both GPU and CPU torch tensors.

**Comment by mingqizhang** [comment]
> @JanuszL Cool! Thanks for doing such an awesome project.
> 
> Just sharing [my code clip](https://gist.github.com/Beanocean/62ba8de2deb5bb2d76474cef2a22bb61) here as an example. I only test the code on Jupyter notebook and the displayed images and masks look OK.
> 
> The correctness on the training procedure still remains to be tested !!!

Sorry, I can not open your code link. Could you please share it again, thanks.

**Comment by Beanocean** [comment]
@mingqizhang It's a [link](https://gist.github.com/Beanocean/62ba8de2deb5bb2d76474cef2a22bb61) from the Gist. I thinks it has open access because some one had forked the code.

**Comment by banasraf** [comment]
@Beanocean 
Hello, 
I have some news about the python operator on GPU. There's a new operator added to DALI - DLTensorPythonFunction. The gpu version of it is currently on master and it's coming in 0.16.
It exposes DALI tensors in python as DLTensors, which can be consumed by different frameworks.
You can see the usage of it here: https://github.com/NVIDIA/DALI/blob/master/dali/test/python/test_dltensor_operator.py 
In the case you would like to run some pytorch operations inside the pipeline, you could do something like this:
```python
torch_stream = torch.cuda.Stream()
def process_data(tensor_list):
    with torch.cuda.stream(torch_stream):
        torch_input = [torch.utils.dlpack.from_dlpack(dltensor) for dltensor in tensor_list]
        torch_output = [some_torch_operation(t) for t in torch_input]
        output = [torch.utils.dlpack.to_dlpack(t) for t in torch_output]
    torch_stream.synchronize()
    return output
``` 
and simply instantiate the operator in the pipeline with: 
```python
self.torch_op = ops.DLTensorPythonFunction(function=process_data, device='gpu')
```
_Note: in this case operator gets the whole batch as an argument, and not single tensors like the PythonFunction op._

The API is still not the most convenient one, but it works, and I'll be working to polish it.

**Comment by Fangyh09** [comment]
@banasraf 
Hello,   torch_stream = torch.cuda.Stream() is a local variable or a global variable?

**Comment by banasraf** [comment]
> @banasraf
> Hello, torch_stream = torch.cuda.Stream() is a local variable or a global variable?

It is global in this case. It has to exist during the execution of the function. Theoretically, we could create a new local stream when the function is called, but it would be more expensive.

**Comment by cceyda** [comment]
> @Beanocean

> ```python
> torch_stream = torch.cuda.Stream()
> def process_data(tensor_list):
>     with torch.cuda.stream(torch_stream):
>         torch_input = [torch.utils.dlpack.from_dlpack(dltensor) for dltensor in tensor_list]
>         torch_output = [some_torch_operation(t) for t in torch_input]
>         output = [torch.utils.dlpack.to_dlpack(t) for t in torch_output]
>     torch_stream.synchronize()
>     return output
> ```

Is the torch_stream stuff in the code doing the same thing as setting the `synchronize_stream` argument [here]( https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html?highlight=stream#nvidia.dali.ops.DLTensorPythonFunction) to True?

**Comment by banasraf** [comment]
@cceyda 
> Is the torch_stream stuff in the code doing the same thing as setting the `synchronize_stream` argument [here](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html?highlight=stream#nvidia.dali.ops.DLTensorPythonFunction) to True?

No. `synchronize_stream` is used to synchronize internal DALI stream before calling the Python function. The `torch_stream.synchronize()` provides synchronization of all the work scheduled inside the Python function.
It looks like this:
```
                             |----torch_stream----------->|
    synchronize_stream=True >|                            |< torch_stream.synchronize()
                             |                            |
-------DALI stream---------->|                            |-----consecutive DALI work-------------> 
```
It's a general rule that if we want to schedule work on a separate stream we should always have those two synchronization points.
Also, there are convenience operators `PythonFunction` and `TorchPythonFunction` available that handle streams and synchronization internally.

**Comment by JanuszL** [comment]
Hi,
Now [`crop` operator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.fn.crop) supports padding thanks to `out_of_bounds_policy`.
To do what you want with resize please use [`peek_image_shape`](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.fn.peek_image_shape) or [`shapes`](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.fn.shapes) operators and then [`uniform`](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.fn.shapes) with the set of predefined values to obtain the desired shape of the image after resize.
